internal/log/zap/zap.go

```diff
@@ -0,0 +1,189 @@
+//
+// Copyright (C) 2019-2021 vdaas.org vald team <vald@vdaas.org>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+package zap
+
+import (
+	"github.com/vdaas/vald/internal/log/format"
+	"github.com/vdaas/vald/internal/log/level"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+const (
+	detailsKey = "details"
+
+	defaultLevel = zapcore.DebugLevel
+)
+
+var (
+	zapcore_NewConsoleEncoder = zapcore.NewConsoleEncoder
+	zapcore_NewJSONEncoder    = zapcore.NewJSONEncoder
+)
+
+type logger struct {
+	format format.Format
+	level  level.Level
+
+	enableCaller bool
+
+	logger *zap.Logger
+	sugar  *zap.SugaredLogger
+}
+
+// New returns a new logger instance.
+func New(opts ...Option) (*logger, error) {
+	l := new(logger)
+	for _, opt := range append(defaultOpts, opts...) {
+		opt(l)
+	}
+
+	err := l.initialize("stdout", "stderr")
+	if err != nil {
+		return nil, err
+	}
+
+	return l, nil
+}
+
+func (l *logger) initialize(sinkPath, errSinkPath string) (err error) {
+	sink, closeOut, err := zap.Open(sinkPath)
+	if err != nil {
+		return err
+	}
+
+	errSink, _, err := zap.Open(errSinkPath)
+	if err != nil {
+		closeOut()
+		return err
+	}
+
+	core := zapcore.NewCore(
+		toZapEncoder(l.format),
+		sink,
+		toZapLevel(l.level),
+	)
+
+	opts := []zap.Option{
+		zap.ErrorOutput(errSink),
+	}
+
+	if l.enableCaller {
+		opts = append(opts, zap.AddCaller())
+	}
+
+	l.logger = zap.New(core, opts...)
+
+	l.sugar = l.logger.Sugar()
```

Comment:
I think we should be define finelizer for logger & sugar.
zap has internal buffering we should call .Sync() method when application exits

Reply:
FYI: https://pkg.go.dev/go.uber.org/zap#Logger.Sync
https://pkg.go.dev/go.uber.org/zap#SugaredLogger.Sync